### PR TITLE
add TPU node pool config

### PR DESCRIPTION
**Pull request by yiyinglovecoding**
add TPU node pool config 
when create TPU node pool
- has to be standard 
- GPU node pool will not be created 

add kuberay-tpu-values.yaml
- worker pod deployed on TPU node as we only have 1 tpu node

### CODE REVIEW

File: ray-on-gke/platform/modules/gke_standard/main.tf [hunk 1]
```diff
@@ -87,3 +96,22 @@ resource "google_container_node_pool" "gpu_pool" {
     }
   }
 }
+
+resource "google_container_node_pool" "multi_host_tpu_pool" {
```
Comment (richardsliu) [inline ray-on-gke/platform/modules/gke_standard/main.tf:100]: The topology is a 2x2x1 which is actually a single host.

Comment (yiyinglovecoding) [reply ray-on-gke/platform/modules/gke_standard/main.tf:100]: modified

File: ray-on-gke/platform/modules/gke_standard/variables.tf [hunk 2]
```diff
@@ -24,6 +24,12 @@ variable "region" {
   default     = "us-central1"
 }
 
+variable "zones" {
+  type = list(string)
+  description = "selected zones for TPU nodes"
+  default = ["us-central2-b"]
```
[the remark this answers is not in the document]
Comment (richardsliu) [reply ray-on-gke/platform/modules/gke_standard/variables.tf:30]: I think it's confusing to have both region and zone here, especially when the above variable description says "region or zone". 

Maybe we can just use the one above, and specify that for TPUs only zones are supported?

File: ray-on-gke/platform/modules/gke_standard/main.tf [hunk 3]
```diff
@@ -87,3 +96,22 @@ resource "google_container_node_pool" "gpu_pool" {
     }
   }
 }
+
+resource "google_container_node_pool" "single_host_tpu_pool" {
```
Comment (richardsliu) [inline ray-on-gke/platform/modules/gke_standard/main.tf:100]: Can we just call it "tpu_pool" for simplicity? The user can actually change it to multihost pods if they want.

File: ray-on-gke/platform/modules/kubernetes/variables.tf [hunk 4]
```diff
@@ -35,3 +35,9 @@ variable "enable_autopilot" {
   description = "Set to true to enable GKE Autopilot clusters"
   default     = false
 }
+
+variable "enable_tpu" {
+  type = bool
+  description = "Set to true to create TPU node pool"
```
Comment (richardsliu) [inline ray-on-gke/platform/modules/kubernetes/variables.tf:41]: Perhaps we should explain here that only one accelerator pool (GPU or TPU) will be created. Does Terraform support an enum value type?

File: ray-on-gke/platform/modules/gke_standard/variables.tf (hunk 2 above)
[the remark this answers is not in the document]
Comment (yiyinglovecoding) [reply ray-on-gke/platform/modules/gke_standard/variables.tf]: removed

**Review by richardsliu — APPROVED** [review]

File: ray-on-gke/user/modules/kuberay/kuberay-tpu-pod-values.yaml [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+# Declare variables to be passed into your templates.
+
+# The KubeRay community welcomes PRs to expose additional configuration
+# in this Helm chart.
+
+image:
+  # Replace this with your own image if needed.
+  # repository: rayproject/ray
+  # tag: 2.3.0-py310
+  repository: gcr.io/yiyingzhang-gke-dev/ray-tpu
```
Comment (richardsliu) [inline ray-on-gke/user/modules/kuberay/kuberay-tpu-pod-values.yaml:26]: We should use the default Ray image here.

File: ray-on-gke/user/modules/kuberay/config/service.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: Service
```
Comment (richardsliu) [inline ray-on-gke/user/modules/kuberay/config/service.yaml:2]: I think this file is not needed.

File: ray-on-gke/user/modules/kuberay/kuberay-tpu-pod-values.yaml [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+      cpu: "1"
+      google.com/tpu: "4"
+      memory: "20G"
+      ephemeral-storage: 10Gi
+  annotations:
+    key: value
+  nodeSelector:
+    #iam.gke.io/gke-metadata-server-enabled: "true"
+    cloud.google.com/gke-tpu-accelerator: tpu-v4-podslice
+    cloud.google.com/gke-tpu-topology: 2x2x2
```
Comment (richardsliu) [inline ray-on-gke/user/modules/kuberay/kuberay-tpu-pod-values.yaml:210]: Can we try with 2x2x1?

File: ray-on-gke/user/modules/kuberay/kuberay-tpu-pod-values.yaml [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+    #     memory: "512Mi"
+    #   requests:
+    #     cpu: "500m"
+    #     memory: "512Mi"
+  labels:
+    cloud.google.com/gke-ray-node-type: head
+  rayStartParams:
+    dashboard-host: '0.0.0.0'
+    block: 'true'
+    resources: '"{\"TPUPOD-$(POD_NAME)\": 1,  \"google.com/tpu\": 4}"'
```
Comment (richardsliu) [inline ray-on-gke/user/modules/kuberay/kuberay-tpu-pod-values.yaml:64]: Remove the first part here (\"TPUPOD-$(POD_NAME)\": 1), it's just for manual testing.

File: ray-on-gke/user/modules/kuberay/kuberay.tf [hunk 9]
```diff
@@ -12,12 +12,21 @@
...
+}
+
 resource "helm_release" "ray-cluster" {
   name       = "example-cluster"
   repository = "https://ray-project.github.io/kuberay-helm/"
   chart      = "ray-cluster"
   namespace  = var.namespace
   values = [
-    file("${path.module}/kuberay-values.yaml")
+    file("${path.module}/kuberay-tpu-pod-values.yaml")
```
Comment (richardsliu) [inline ray-on-gke/user/modules/kuberay/kuberay.tf:30]: I think it is better to use kuberay-tpu-values.yaml as a starting point.

File: ray-on-gke/user/modules/kuberay/kuberay-tpu-values.yaml [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+  # we don't recommend allocating less than 8G memory for a Ray pod in production.
+  # Ray pods should be sized to take up entire K8s nodes when possible.
+  # Always set CPU and memory limits for Ray pods.
+  # It is usually best to set requests equal to limits.
+  # See https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/config.html#resources
+  # for further guidance.
+  resources:
+    limits:
+      cpu: "8"
+      # google.com/tpu: "4"
```
Comment (richardsliu) [inline ray-on-gke/user/modules/kuberay/kuberay-tpu-values.yaml:96]: Remove this line if not needed.

File: ray-on-gke/user/modules/kuberay/kuberay-tpu-values.yaml [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+      # To avoid out-of-memory issues, never allocate less than 2G memory for the Ray head.
+      memory: "40G"
+      ephemeral-storage: 20Gi
+    requests:
+      cpu: "8"
+      # google.com/tpu: "4"
```
Comment (richardsliu) [inline ray-on-gke/user/modules/kuberay/kuberay-tpu-values.yaml:102]: Here as well.

**Review by richardsliu — APPROVED** [review]

**Review by richardsliu — APPROVED** [review]

**Pull request merged**